src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs

```diff
@@ -0,0 +1,66 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+// NOTE: This file is shared between CoreFX and ASP.NET.  Be very thoughtful when changing it.
+
+namespace System.Net.WebSockets
+{
+    internal sealed partial class ManagedWebSocket : WebSocket
+    {
+        public override Task SendAsync(ReadOnlyMemory<byte> buffer, WebSocketMessageType messageType, bool endOfMessage, CancellationToken cancellationToken)
+        {
+            return SendPrivateAsync(buffer, messageType, endOfMessage, cancellationToken);
+        }
+
+        public override ValueTask<ValueWebSocketReceiveResult> ReceiveAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            try
+            {
+                WebSocketValidate.ThrowIfInvalidState(_state, _disposed, s_validReceiveStates);
+
+                Debug.Assert(!Monitor.IsEntered(StateUpdateLock), $"{nameof(StateUpdateLock)} must never be held when acquiring {nameof(ReceiveAsyncLock)}");
+                lock (ReceiveAsyncLock) // synchronize with receives in CloseAsync
+                {
+                    ThrowIfOperationInProgress(_lastReceiveAsync);
+                    ValueTask<ValueWebSocketReceiveResult> t = ReceiveAsyncPrivate<ValueWebSocketReceiveResultGetter, ValueWebSocketReceiveResult>(buffer, cancellationToken);
+                    _lastReceiveAsync =
+                        t.IsCompletedSuccessfully ? (t.Result.MessageType == WebSocketMessageType.Close ? s_cachedCloseTask : Task.CompletedTask) :
+                        t.AsTask();
+                    return t;
+                }
+            }
+            catch (Exception exc)
+            {
+                return new ValueTask<ValueWebSocketReceiveResult>(Task.FromException<ValueWebSocketReceiveResult>(exc));
+            }
+        }
+
+        private Task ValidateAndReceiveAsync(Task receiveTask, byte[] buffer, CancellationToken cancellationToken)
+        {
+            if (receiveTask == null ||
+                        (receiveTask.Status == TaskStatus.RanToCompletion &&
```

Comment:
This can be changed from:
```C#
receiveTask.Status == TaskStatus.RanToCompletion
```
to
```C#
receiveTask.IsCompletedSuccessfully
```
They're functionally identical but the latter is more efficient.

Reply:
Looking at the property implementation, it looks like it is doing the same check, how is it efficient?

Reply:
What implementations are you comparing?

Reply:
Was looking at this, https://github.com/dotnet/corefx/blob/master/src/System.Threading.Tasks.Extensions/src/System/Threading/Tasks/ValueTask.cs#L118

Reply:
That's ValueTask.  This is Task.

Reply:
And that's also the corefx implementation of ValueTask, which works downlevel.  Compare it to the coreclr implementation: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Threading/Tasks/ValueTask.cs#L132

Reply:
I still don't get why it's more efficient though..

Reply:
> I still don't get why it's more efficient though

Have you looked at the implementations?
https://github.com/dotnet/coreclr/blob/140ba024bf6edb61744fe6b84bd635d389fe1413/src/mscorlib/src/System/Threading/Tasks/Task.cs#L1458
vs
https://github.com/dotnet/coreclr/blob/140ba024bf6edb61744fe6b84bd635d389fe1413/src/mscorlib/src/System/Threading/Tasks/Task.cs#L1329-L1351